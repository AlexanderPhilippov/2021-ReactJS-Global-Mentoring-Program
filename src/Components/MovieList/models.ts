import { MovieModel } from 'Components/Movie/models'
import { MovieFormAction } from 'Components/MovieForm/models'

export interface MoviesResponseModel {
    data: MovieModel[]
    totalAmount: number
    offset: number
    limit: number
}

export interface MoviesHeaderProps {
    total?: number
}

export interface MovieListLocalState {
    isModalOpen: boolean
    selectedMovie?: MovieModel
    currentAction?: MovieFormAction
}

export interface MovieListReducerAction {
    payload: MovieListLocalState
}

export interface MovieListState extends Partial<MoviesResponseModel> {
    isLoading: boolean
    selectedGenre?: string
    sortBy?: string
    error?: string
}

export interface MovieListAction {
    type: string
    payload?: MoviesResponseModel
    error?: string
}
