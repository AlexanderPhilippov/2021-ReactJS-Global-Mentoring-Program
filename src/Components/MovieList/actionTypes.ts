export default {
    FETCH_MOVIES_BEGIN: 'FETCH_MOVIES_BEGIN',
    FETCH_MOVIES_SUCCESS: 'FETCH_MOVIES_SUCCESS',
    FETCH_MOVIES_ERROR: 'FETCH_MOVIES_ERROR',
    SET_MOVIES_FILTER_GENRE: 'SET_MOVIES_FILTER_GENRE',
    SET_MOVIES_FILTER_SORT_BY: 'SET_MOVIES_FILTER_SORT_BY',
    SET_MOVIES_FILTER_SORT_ORDER: 'SET_MOVIES_FILTER_SORT_ORDER',
    SET_MOVIES_FILTER_SEARCH: 'SET_MOVIES_FILTER_SEARCH',
    SET_MOVIES_FILTER_SEARCH_BY: 'SET_MOVIES_FILTER_SEARCH_BY',
    SET_MOVIES_FILTER_REFRESH_REQUIRED: 'SET_MOVIES_FILTER_REFRESH_REQUIRED',
}
