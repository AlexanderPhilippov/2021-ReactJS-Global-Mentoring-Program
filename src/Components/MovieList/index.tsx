import React, { useState } from 'react'
import Movie from 'Components/Movie'
import { MoviesResponseModel } from './models'
import MoviesMockData from './movies.json'

const MovieList: React.FC = () => {

    const [moviesResponse] = useState<MoviesResponseModel>(MoviesMockData)

    return (
        <>
            {moviesResponse?.data.map(x => (
                <Movie key={x.id} movie={x} />
            ))}
        </>
    )
}

export default MovieList
