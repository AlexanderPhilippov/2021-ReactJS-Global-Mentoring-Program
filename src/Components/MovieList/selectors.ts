import { MovieModel } from 'Components/Movie/models'
import { createSelector } from 'reselect'
import { AppState } from 'src/Store/rootReducer'
import { SearchBy, SortOrder } from './models'

const moviesState = (state: AppState) => state.movies
const filterState = (state: AppState) => state.filter

export const getTotalAmountSelector = createSelector(
    moviesState,
    (movies) => movies?.totalAmount || 0
)

export const getMoviesSelector = createSelector(
    moviesState,
    (movies) => movies?.data || []
)

export const getMovieByIdSelector = (
    state: AppState,
    id?: number
): MovieModel =>
    createSelector(
        getMoviesSelector,
        (movies) => movies.find((movie) => movie.id === id) as MovieModel
    )(state)

export const getErrorSelector = createSelector(
    moviesState,
    (movies) => movies?.error
)

export const getGenreSelector = createSelector(
    filterState,
    (filter) => filter.genre || SearchBy.ALL
)

export const getSotrtBySelector = createSelector(
    filterState,
    (filter) => filter.sortBy || SearchBy.GENRES
)

export const getSortOrderSelector = createSelector(
    filterState,
    (filter) => filter.sortOrder || SortOrder.DESC
)

export const getSearchSelector = createSelector(
    filterState,
    (filter) => filter.search || ''
)

export const getSearchBySelector = createSelector(
    filterState,
    (filter) => filter.searchBy || SearchBy.ALL
)
