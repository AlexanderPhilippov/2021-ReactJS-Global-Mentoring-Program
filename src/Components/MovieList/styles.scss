.movie-list {
    display: flex;
    flex-wrap: wrap;
    flex-direction: row;
    justify-content: space-around;
    align-items: flex-start;
    padding: 0 10%;

    &-header{
        flex: 1 1 100%;
        padding: 25px 2% 0 2%;
        position: relative;
        display: flex;
        align-items: center;
        justify-content: space-between;

        &__genres, &__sorting{
            display: flex;
            color: white;
            font-size: 22px;
            margin-bottom: 20px;
            border-bottom: 1px solid #424242;
        }

        &__genres {
            flex: 1 1 60%;

            div:first-child {
                border-bottom: 1px solid #F65261;

            }

            div {
                margin-right: 15px;
                text-transform: uppercase;
            }
        }
        
        &__sorting {
            div {
                padding-right: 15px;                
                text-transform: uppercase;
            }

            select {
                border: none;
                border-radius: 4px;
                width: 200px;
                background-color: #232323;
                color: white;
                font-size: 18px;
                text-align: center;
                text-transform: uppercase;
            }
        }
    }
}