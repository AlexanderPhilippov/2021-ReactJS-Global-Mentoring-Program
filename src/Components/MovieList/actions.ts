import actionTypes from './actionTypes'
import {
    MovieListAction,
    MovieListFilterAction,
    MoviesResponseModel,
} from './models'

export const fetchMoviesBegin = (): MovieListAction => ({
    type: actionTypes.FETCH_MOVIES_BEGIN,
})

export const fetchMoviesSuccess = (
    data: MoviesResponseModel
): MovieListAction => ({
    type: actionTypes.FETCH_MOVIES_SUCCESS,
    payload: data,
})

export const fetchMoviesError = (data: string): MovieListAction => ({
    type: actionTypes.FETCH_MOVIES_ERROR,
    error: data,
})

export const setMoviesGenre = (data: string): MovieListFilterAction => ({
    type: actionTypes.SET_MOVIES_FILTER_GENRE,
    payload: { genre: data },
})
