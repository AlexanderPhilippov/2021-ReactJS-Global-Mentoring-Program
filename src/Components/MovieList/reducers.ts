import actionTypes from './actionTypes'
import {
    MovieListState,
    MovieListAction,
    MovieListFilterState,
    MovieListFilterAction,
    SortOrder,
    SearchBy,
} from './models'

export const movieListReducer = (
    state: MovieListState = { isLoading: false },
    action: MovieListAction
): MovieListState => {
    switch (action.type) {
        case actionTypes.FETCH_MOVIES_BEGIN:
            return { ...state, isLoading: true }
        case actionTypes.FETCH_MOVIES_SUCCESS:
            return { ...action.payload, isLoading: false }
        case actionTypes.FETCH_MOVIES_ERROR:
            return { error: action.error, isLoading: false }
        default:
            return state
    }
}

export const movieListFilterReducer = (
    state: MovieListFilterState = {
        genre: SearchBy.ALL,
        sortBy: SearchBy.GENRES,
        sortOrder: SortOrder.DESC,
        search: '',
        searchBy: SearchBy.TITLE,
    },
    action: MovieListFilterAction
): MovieListFilterState => {
    switch (action.type) {
        case actionTypes.SET_MOVIES_FILTER_GENRE:
            return { ...state, genre: action.payload.genre }
        case actionTypes.SET_MOVIES_FILTER_SORT_BY:
            return { ...state, sortBy: action.payload.sortBy }
        case actionTypes.SET_MOVIES_FILTER_SORT_ORDER:
            return { ...state, sortOrder: action.payload.sortOrder }
        case actionTypes.SET_MOVIES_FILTER_SEARCH:
            return { ...state, search: action.payload.search }
        case actionTypes.SET_MOVIES_FILTER_SEARCH_BY:
            return { ...state, searchBy: action.payload.searchBy }
        case actionTypes.SET_MOVIES_FILTER_REFRESH_REQUIRED:
            return { ...state, refreshRequired: action.payload.refreshRequired }
        default:
            return state
    }
}
