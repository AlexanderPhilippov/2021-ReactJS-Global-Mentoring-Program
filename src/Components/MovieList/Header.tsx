import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import classnames from 'classnames'
import { SortOrder } from './models'
import { setMoviesGenre, setSortBy, setSortOrder } from './actions'
import * as Selectors from './selectors'

const GenreLink: React.FC<{
    name: string
    isSelected?: boolean
}> = ({ name, isSelected }) => {
    const dispatch = useDispatch()
    const handleClick = () => {
        dispatch(setMoviesGenre(name))
    }
    return (
        <div
            className={classnames('movie-list-header__genre-link', {
                active: isSelected,
            })}
            onClick={handleClick}
        >
            {name || 'all'}
        </div>
    )
}
const MoviesHeader: React.FC = () => {
    const genres = ['', 'documentary', 'comedy', 'horror', 'crime']

    const sortOrder = useSelector(Selectors.getSortOrderSelector)
    const genre = useSelector(Selectors.getGenreSelector)
    const totalAmount = useSelector(Selectors.getTotalAmountSelector)

    const dispatch = useDispatch()

    const handleChange = (e: React.FormEvent<HTMLSelectElement>) => {
        dispatch(setSortBy(e.currentTarget.value))
    }

    const handleChangeSortOrder = () => {
        dispatch(
            setSortOrder(
                sortOrder === SortOrder.DESC ? SortOrder.ASC : SortOrder.DESC
            )
        )
    }
    return (
        <div className="movie-list-header">
            <div className="movie-list-header__genres">
                {genres.map((genreName) => {
                    return (
                        <GenreLink
                            key={genreName}
                            name={genreName}
                            isSelected={genreName === genre}
                        />
                    )
                })}
            </div>
            <div className="movie-list-header__sorting">
                <div className="">
                    sort by
                    <span
                        onClick={handleChangeSortOrder}
                        className="movie-list-header__sorting_order"
                    >
                        ({sortOrder})
                    </span>
                </div>
                <select onChange={handleChange}>
                    <option value="genres">genre</option>
                    <option value="vote_average">rating</option>
                    <option value="release_date">release date</option>
                </select>
            </div>
            <div className="movie-list-header__total">
                <span>{totalAmount}</span>&nbsp;movies found
            </div>
        </div>
    )
}

export { MoviesHeader }
