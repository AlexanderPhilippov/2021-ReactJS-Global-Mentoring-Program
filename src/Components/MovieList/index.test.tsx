/**
 * @jest-environment jsdom
 */

import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import MovieList from './index'
import { Router, MemoryRouter } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import { SortOrder } from './models'

const mockStore = configureStore()
const mockDispatch = jest.fn()

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useDispatch: () => mockDispatch,
}))

jest.mock('./selectors.ts', () => ({
    getRefreshRequiredFlagSelector: () => '',
    getMoviesSelector: () => [],
    getTotalAmountSelector: () => 0,
}))

jest.mock('Components/Modal/selectors.ts', () => ({
    isModalOpen: () => false,
}))

const mockedPush = jest.fn()
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockedPush,
    }),
}))

describe('Interactive components tests', () => {
    it('Snapshot', () => {
        const store = mockStore({})
        const { asFragment } = render(
            <Provider store={store}>
                <MemoryRouter>
                    <MovieList />
                </MemoryRouter>
            </Provider>
        )
        expect(asFragment()).toMatchSnapshot()
    })
    it('Genre select', () => {
        const history = createMemoryHistory()
        const store = mockStore({})
        const { getByText } = render(
            <Provider store={store}>
                <Router history={history}>
                    <MovieList />
                </Router>
            </Provider>
        )
        expect(history.location.pathname).toBe('/')

        fireEvent.click(getByText(/Documentary/i))
        expect(history.location.pathname).toBe('/search/Documentary/')

        fireEvent.click(getByText(/Drama/i))
        expect(history.location.pathname).toBe('/search/Drama/')

        fireEvent.click(getByText(/Animation/i))
        expect(history.location.pathname).toBe('/search/Animation/')
    })

    it('Sort order', () => {
        const store = mockStore({})
        const { getByText } = render(
            <Provider store={store}>
                <MemoryRouter>
                    <MovieList />
                </MemoryRouter>
            </Provider>
        )
        fireEvent.click(getByText(`(${SortOrder.DESC})`))
        expect(mockedPush).toHaveBeenCalledTimes(1)
        expect(mockedPush).toHaveBeenCalledWith({
            pathname: '/',
            search: `sortOrder=${SortOrder.ASC}`,
        })
    })
})
