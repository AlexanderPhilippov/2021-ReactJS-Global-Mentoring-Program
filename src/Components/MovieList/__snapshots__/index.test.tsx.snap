// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interactive components tests Snapshot 1`] = `
<DocumentFragment>
  <div
    class="movie-list"
  >
    <div
      class="movie-list-header"
    >
      <div
        class="movie-list-header__genres"
      >
        <a
          aria-current="page"
          class="movie-list-header__genre-link active"
          href="/search/All/"
        >
          All
        </a>
        <a
          class="movie-list-header__genre-link"
          href="/search/Documentary/"
        >
          Documentary
        </a>
        <a
          class="movie-list-header__genre-link"
          href="/search/Comedy/"
        >
          Comedy
        </a>
        <a
          class="movie-list-header__genre-link"
          href="/search/Horror/"
        >
          Horror
        </a>
        <a
          class="movie-list-header__genre-link"
          href="/search/Crime/"
        >
          Crime
        </a>
        <a
          class="movie-list-header__genre-link"
          href="/search/Drama/"
        >
          Drama
        </a>
        <a
          class="movie-list-header__genre-link"
          href="/search/Animation/"
        >
          Animation
        </a>
      </div>
      <div
        class="movie-list-header__sorting"
      >
        <div
          class=""
        >
          sort by
          <span
            class="movie-list-header__sorting_order"
          >
            (desc)
          </span>
        </div>
        <select>
          <option
            selected=""
            value="genres"
          >
            genre
          </option>
          <option
            value="vote_average"
          >
            rating
          </option>
          <option
            value="release_date"
          >
            release date
          </option>
        </select>
      </div>
      <div
        class="movie-list-header__total"
      >
        <span>
          0
        </span>
        Â movies found
      </div>
    </div>
  </div>
</DocumentFragment>
`;
